flow/aggregates.py

```diff
@@ -0,0 +1,498 @@
+    """
+    def __init__(self, project):
+        self._project = project
+
+    def __iter__(self):
+        for job in self._project:
+            yield (job,)
+
+    def __getitem__(self, id):
+        "Return a tuple of a single job via job id."
+        try:
+            return (self._project.open_job(id=id),)
+        except KeyError:
+            raise LookupError(f"Did not find job with id {id}.")
+
+    def __contains__(self, job):
+        """Return whether the job is present in the project associated with this
+        instance of :py:class:`_DefaultAggregateStore`.
+        """
+        return job in self._project
+
+    def __len__(self):
+        return len(self._project)
+
+    def __eq__(self, other):
+        return type(self) == type(other) and \
+               self._project == other._project
+
+    def __hash__(self):
+        blob = self._project.id
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _register_aggregates(self, project):
+        """We have to store self._project when this method is invoked
+        This is because we will then iterate over that project in
+        order to return an aggregates of one.
+        """
+        self._project = project
+
+
+class _MakeAggregates:
+    """This class acts as a funnctor which handles the creation of
+    aggregates using all the jobs in a project.
+
+    :param aggregator_function:
+        The ``aggregator_function`` attribute associated with the
+        :py:class:`aggregator`
+    :type aggregator_function:
+        callable
+    :param sort_by:
+        The ``sort_by`` attribute associated with the :py:class:`aggregator`
+    :type sort_by:
+        str or NoneType
+    :param reverse_order:
+        The ``reverse_order`` attribute associated with the :py:class:`aggregator`
+    :type reverse_order:
+        bool
+    :param select:
+        The ``select`` attribute associated with the :py:class:`aggregator`
+    :type select:
+        callable or NoneType
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, aggregator_function, sort_by, reverse_order, select, project):
+        self._aggregator_function = aggregator_function
+        self._sort_by = sort_by
+        self._reverse_order = bool(reverse_order)
+        self._select = select
+        self._project = project
+
+    def __call__(self):
+        jobs = list(self._project)
+
+        if self._select is not None:
+            jobs = list(filter(self._select, jobs))
+        if self._sort_by is not None:
+            jobs = sorted(jobs,
+                          key=lambda job: job.sp[self._sort_by],
+                          reverse=self._reverse_order)
+        aggregated_jobs = self._aggregator_function(jobs)
+        return self._create_nested_aggregate_list(aggregated_jobs)
```

Comment:
This should still be done by the `_AggregateStore` class. I believe the `_MakeAggregate` class should only be responsible for correctly handing the aggregation, not the data structures used by `_AggregateStore`.